src/backup.py

```diff
@@ -0,0 +1,133 @@
+        region: The region to connect to the object storage.
+        bucket: The bucket name.
+        endpoint: The endpoint used to connect to the object storage.
+        path: The path inside the bucket to store objects.
+        s3_uri_style: The S3 protocol specific bucket path lookup type.
+    """
+
+    access_key: str = Field(alias="access-key")
+    secret_key: str = Field(alias="secret-key")
+    region: Optional[str] = Field(default=None)
+    bucket: str
+    endpoint: Optional[str]
+    path: str = Field(default="")
+    s3_uri_style: str = Field(alias="s3-uri-style", default="host")
+
+    @validator("region", always=True)
+    @classmethod
+    def check_region_or_endpoint_set(cls, region: str, values: dict[str, Any]) -> str:
+        """Validate that eather that region or endpoint are set.
+
+        Args:
+            region: region attribute
+            values: all attributes in S3 configuration
+
+        Returns:
+            value of the region attribute
+
+        Raises:
+            ValueError: if the configuration is invalid.
+        """
+        endpoint = values.get("endpoint")
+        if not region and not endpoint:
+            raise ValueError('one of "region" or "endpoint" needs to be set')
+        return region
+
+
+class SynapseBackup(Object):
+    """Class to manage Synapse backups over S3."""
+
+    _S3_RELATION_NAME = "s3-backup-parameters"
+
+    def __init__(self, charm: ops.CharmBase):
+        """Initialize the backup object.
+
+        Args:
+            charm: The parent charm the backups are made for.
+        """
+        super().__init__(charm, "backup")
+
+        self._charm = charm
+        self._s3_client = S3Requirer(self._charm, self._S3_RELATION_NAME)
+        self.framework.observe(
+            self._s3_client.on.credentials_changed, self._on_s3_credential_changed
+        )
+        self.framework.observe(self._s3_client.on.credentials_gone, self._on_s3_credential_gone)
+
+    def _on_s3_credential_changed(self, _: CredentialsChangedEvent) -> None:
+        """Check new S3 credentials set the unit to blocked if they are wrong."""
+        try:
+            s3_parameters = S3Parameters(**self._s3_client.get_s3_connection_info())
+        except ValueError:
+            self._charm.unit.status = ops.BlockedStatus(S3_INVALID_CONFIGURATION)
+            return
+
+        if not can_use_bucket(s3_parameters):
+            self._charm.unit.status = ops.BlockedStatus(S3_CANNOT_ACCESS_BUCKET)
+        logger.info("S3 backup correctly configured")
+
+    def _on_s3_credential_gone(self, _: CredentialsChangedEvent) -> None:
+        """Handle s3 credentials gone to reset unit status if it is now correct."""
+        if (
+            isinstance(self._charm.unit.status, ops.BlockedStatus)
+            and self._charm.unit.status.message in BACK_UP_STATUS_MESSAGES
+        ):
+            self._charm.unit.status = ops.ActiveStatus()
+
+
+def can_use_bucket(s3_parameters: S3Parameters) -> bool:
```

Comment:
To be compliant with our [contributing guide](https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#function-and-method-Ordering) , the function should appear before it is used, so above `_on_s3_credential_changed`.